### DISCUSSION TITLE
Feature Requests

### DISCUSSION THREAD

**Discussion by LucasFaudman** [LucasFaudman/linkedin-automator#2] [Ideas]
# Request Features for LinkedIn AI

Describe new features you'd like to see. All requests welcome.

Features to be added:
- Contacts DB for viewing Hiring Managers and other Contacts
- Automated Connection Requests
- Automated Custom Messages
- AI Job Fit Evaluation by @nikdaya

**Comment by LucasFaudman** [comment]
By @nikdaya originally in #1 

Would you consider using AI to assign a rate 0 to 100 regarding the most matching job, given as input to the AI Assistant the title and description of the job along with the skills required, and the resume.txt?
Something like:
def evaluate_job(resume, title, description, skills):
    # give to AI resume, title, description, skills
    # prompt is to give 0 - 100 answer
    # save the score into the "score" column of the jobs.db
    # etc...

This would allow to not only find jobs and apply but also to apply for the topmost matching jobs, evaluated by AI. In other words, LinkedIn suggestions are not exactly "suggestions", more like an "attempt of suggestion" and it would be great to have an AI estimation on these.

**Comment by nikdaya** [comment]
## Feature requests

### Order by column in the job database section
1. I would like to have the possibility to order by any column of the job database by clicking on the header of the column. For example, ordering by `date_applied` would order by date ascending/descending. Ordering by `rate` would order by the rate of the job position outputted by the AI.

### Max daily job applications
2. I would like to have in the configuration file a `max_job_daily_application` parameter where I would choose how many job application to perform daily. For example 100, at the 100th a pop-up window warns me that "Job application daily limit reached (100). Change configuration file to perform more applications today or come back at {date and hour of today + 24h}". There should be also the possibility to set the `max_job_daily_application` to `None` so that the limit is unlimited.

### Button "apply to jobs that satisty rate threshold"
3. In the "Search and apply for jobs" section, I would like to have a button where I could apply for the jobs that have been rated higher than the `rate_job_threshold` parameter, set up in the configuration file. This button applies ONLY to the job applications that rate over this parameter, skipping all the other ones. For example, let's say I have only 5 applications where I rate over 80. Clicking on this button, I apply only to 5 applications.

### Charts tab
4. I would like to have a new tab "Plotting" or similar name where I have some graphical representations of my applications (number of daily applications per day, applications grouped by place, applications grouped by company, number of applications where I received feedback or my CV was viewed, number of applications where I applied but then the state of the application changed to "closed", etc... [to elaborate the most interesting plots furtherly]).

### Telegram API integration (telethon)
5. I would like to have, via the library `telethon` and by defining the API key and API hash in the configuration file, along with the phone number and group name, possibility to be prompted for questions that AI does not know, via a Telegram group instead of only the program interface. I would like to choose via configuration file if to give answers via Telegram, via interface or both. Such classes should handle both sending the message on the channel as well as listening for the user typing for the answer, in an async way. In this way it would be possible, by leaving the program running, answer for the questions from literally anywhere.

**Reply by LucasFaudman** [reply]
@nikdaya 

Thank you for your awesome suggestions, and for reporting the errors in the other thread perfectly. I appreciate it!

I'm currently working on a new library that will replace the current AI backend and allow the use function-calling compatible local models via Ollama _entirely for free_, as well as other paid Anthropic models. 

The next update will have minimal feature changes and will basically be the same as before but with options to switch between AI providers.

After everything is stable again with the new backend, then I will get to these features. I can tell you for sure that I will be adding suggestions 1-3 in the next 2-3 updates and that I will look into 4 and 5 as well. However, I am currently traveling for work so I can't say exactly when I'll release the update yet.  

Thanks again and take care!

**Reply by nikdaya** [reply]
@LucasFaudman

You're welcome Lucas! Glad I could contribute.
This idea of substituting the back-end with a local model (or any model) of choice, it is amazing. Gives me so many ideas where the AI model could be prompted for many other results.

Take your time to dedicate for work and to consolidate the updates regarding the back-end. I'll be putting some other features suggests down here. Wish I could contribute more with code, as far as I'm concerned I have some knowledge in python but not exactly in the libraries you're using (Gtk as an example), if in the future I'll be more skilled in these I'll consider building a PR.

PS: I suggest you to setup a "donate" button on the page for contributions to the library, and also a page for PR submit instructions and issue submit template. Take care!

---

## Search "broad best fit" functionality
6. Sometimes we're just stuck on a city to find a job, and simply it's not the best fit apart from in our imagination. What about building a model that searches for us "the best match" based on some acquired data? For example, let's say we put US as a binding for our search. We could use [this](https://livingcost.org/best/united-states) to acquire data about what are the best US cities to live in based on best quality of life index. Then the model will search for all these cities (up to 100 but this could be parametrized), get a list of jobs for each, for each calculate the score (see feature request #0 at the start of the thread) and the final score will be the [geometric median](https://en.wikipedia.org/wiki/Geometric_median) of the score and the quality of life score (square root of the multiplication of these two scores). Served then to the user the matches that maximize this median, by over the `median_rate_job_threshold` parameter or similar. So let's say only the jobs that meet a score of 80 or more.

**Comment by nikdaya** [comment]
Hey @LucasFaudman, I was wondering how is the testing of the different back-ends going on?
I saw no other commits on main so I guess the testing is still on.

**Reply by LucasFaudman** [reply]
Hey @nikdaya, 

Thanks again for your inputs and for checking in! 

Development on the new unified AI client library has been going really well, however it's turned into a much bigger undertaking than I realized when I first mentioned it to you. What started as just a component for this project and a few others has grown into a full library that I've been working hard to make absolutely perfect since it's going to replace the OpenAI cloud function infra for a start-up I've been working for and I'm planning to release it publicly soon.

Some updates:
- Chat and standard Tool-Calling features now fully supported for all capable models from providers:
    - Anthropic
    - GoogleAI/Gemini
    - Ollama
    - OpenAI
- Advanced Tool-Calling upgrades fullying working for all providers except Ollama. (Some issues with the Ollama's modelfile handling of Tools that have yet to be fixed are making this difficult) 
- Image support working for all except OpenAI (OpenAI deprecated `gpt-4-vision-preview` and as of 9/17 have no chat completion model endpoints that can accept image inputs)
- Embeddings, Vector DB, and Auto-RAG support in progress (Retrieval-Augmented Generation: using `Vector Similarity` to determine the most relevant info to provide a model to help it handle a given prompt)
- Finally picked a name I like: `UnifAI` since the goal is to Unify the syntax and implementation of AI tools across all providers.

The repo is still private for now but will be available here soon: 
https://github.com/LucasFaudman/UnifAI 
https://pypi.org/project/unifai/

**Reply by LucasFaudman** [reply]
Regarding updates to `linkedin-ai`, don't worry, I will be adding _all_ the features you suggested as part of testing `UnifAI` once it's stable. 

I'm shooting for next week, but tbh I can only really work on this project in my free time, so I can't promise a date, but I'll be sure to tag you in the commits with your suggestions.

PS: Thanks for suggesting adding a "Donate" button, means a lot. Take care and best of luck with your job search!
